File: src/main/java/ru/serdtsev/homemoney/account/BalanceService.java
package ru.serdtsev.homemoney.account;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.springframework.stereotype.Service;
import ru.serdtsev.homemoney.account.model.AccountType;
import ru.serdtsev.homemoney.account.model.Balance;
import ru.serdtsev.homemoney.balancesheet.BalanceSheet;
import ru.serdtsev.homemoney.balancesheet.BalanceSheetRepository;
import ru.serdtsev.homemoney.moneyoper.MoneyOperItemRepo;
import ru.serdtsev.homemoney.moneyoper.MoneyOperService;

import java.util.Comparator;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
public class BalanceService {
  private final BalanceSheetRepository balanceSheetRepo;
  private final BalanceRepository balanceRepo;
  private final MoneyOperItemRepo moneyOperItemRepo;
  private final ReserveRepository reserveRepo;
  private final MoneyOperService moneyOperService;

  public List<Balance> getBalances(UUID bsId) {
    val balanceSheet = balanceSheetRepo.findOne(bsId);
    return ((List<Balance>) balanceRepo.findByBalanceSheet(balanceSheet)).stream()
        .filter(balance -> balance.getType() != AccountType.reserve)
        .sorted(Comparator.comparing(Balance::getNum))
        .collect(Collectors.toList());
  }

  public void createBalance(UUID bsId, UUID balanceId) {
    val balanceSheet = balanceSheetRepo.findOne(bsId);
    val balance = balanceRepo.findOne(balanceId);
    balance.setBalanceSheet(balanceSheet);
    balance.init(reserveRepo);
    balanceRepo.save(balance);
  }

  public void updateBalance(Balance balance) {
    val storedBalance = balanceRepo.findOne(balance.getId());
    storedBalance.merge(balance, reserveRepo, moneyOperService);
    balanceRepo.save(storedBalance);
  }

  public void deleteOrArchiveBalance(UUID balanceId) {
    val balance = balanceRepo.findOne(balanceId);
    val operFound = moneyOperItemRepo.findByBalance(balance).limit(1).count() > 0;
    if (operFound) {
      balance.setArc(true);
      balanceRepo.save(balance);
      log.info("{} moved to archive.", balance);
    } else {
      balanceRepo.delete(balance);
      log.info("{} deleted.", balance);
    }
  }

  public void upBalance(UUID bsId, UUID balanceId) {
    // todo работает неправильно, исправить
    BalanceSheet bs = balanceSheetRepo.findOne(bsId);
    Balance balance = balanceRepo.findOne(balanceId);

    List<Balance> balances = bs.getBalances().stream()
        .sorted((b1, b2) -> b1.getNum() < b2.getNum() ? -1 : (b1.getNum() > b2.getNum() ? 1 : 0))
        .collect(Collectors.toList());
    assert !balances.isEmpty();

    Balance prev;
    do {
      int index = balances.indexOf(balance);
      assert index > -1;
      prev = null;
      if (index > 0) {
        prev = balances.get(index - 1);
        balances.set(index - 1, balance);
        balances.set(index, prev);
        long i = 0;
        for (Balance b : balances) {
          b.setNum(i++);
        }
      }
    } while (prev != null && prev.getArc());
    balanceRepo.save(balances);
  }

}

File: src/main/kotlin/ru/serdtsev/homemoney/account/BalanceService.kt
package ru.serdtsev.homemoney.account

import mu.KotlinLogging
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import ru.serdtsev.homemoney.account.model.AccountType
import ru.serdtsev.homemoney.account.model.Balance
import ru.serdtsev.homemoney.common.ApiRequestContextHolder
import ru.serdtsev.homemoney.moneyoper.MoneyOperService
import ru.serdtsev.homemoney.moneyoper.dao.MoneyOperItemRepo
import java.util.*

@Service
class BalanceService(
    private val apiRequestContextHolder: ApiRequestContextHolder,
    private val balanceRepo: BalanceRepository,
    private val moneyOperItemRepo: MoneyOperItemRepo,
    private val reserveRepo: ReserveRepository,
    private val moneyOperService: MoneyOperService
) {
    fun getBalances(): List<Balance> {
        val balanceSheet = apiRequestContextHolder.getBalanceSheet()
        return balanceRepo.findByBalanceSheet(balanceSheet)
            .filter { it.type != AccountType.reserve }
            .sortedBy { it.num }
    }

    fun createBalance(balance: Balance) {
        balance.init(reserveRepo)
        balanceRepo.save(balance)
    }

    fun updateBalance(balance: Balance) {
        val storedBalance = balanceRepo.findByIdOrNull(balance.id)!!
        storedBalance.merge(balance, reserveRepo, moneyOperService)
        balanceRepo.save(storedBalance)
    }

    fun deleteOrArchiveBalance(balanceId: UUID) {
        val balance = balanceRepo.findByIdOrNull(balanceId)!!
        val operFound = moneyOperItemRepo.findByBalance(balance).take(1).count() > 0
        if (operFound) {
            balance.isArc = true
            balanceRepo.save(balance)
            log.info { "$balance moved to archive." }
        } else {
            balanceRepo.delete(balance)
            log.info { "$balance deleted." }
        }
    }

    fun upBalance(balanceId: UUID) {
        // todo работает неправильно, исправить
        val balanceSheet = apiRequestContextHolder.getBalanceSheet()
        val balance = balanceRepo.findByIdOrNull(balanceId)!!
        val balances = balanceSheet.balances.sortedBy { it.num }.toMutableList()
        assert(balances.isNotEmpty())
        var prev: Balance?
        do {
            val index = balances.indexOf(balance)
            assert(index > -1)
            prev = null
            if (index > 0) {
                prev = balances[index - 1]
                balances[index - 1] = balance
                balances[index] = prev
                var i: Long = 0
                for (b in balances) {
                    b.num = i++
                }
            }
        } while (prev != null && prev.isArc!!)
        balances.forEach { balanceRepo.save(it) }
    }

    companion object {
        private val log = KotlinLogging.logger {  }
    }
}